LEGO DATA MANIPULATION

![LegoSchema](Lego_Schema.PNG)

## Overview

This project includes various tables from the [Rebrickable](https://rebrickable.com/) website. Rebrickable is a fantastic website that is used to show LEGO sets that are buildable from existing sets.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above. [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation. 


In [2]:
# common imports
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

## Import Data


In [3]:
# import Lego data sets here
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

In [4]:
colors.head()

,id,name,rgb,is_trans
0,-1,[Unknown],0033B2,f
1,0,Black,05131D,f
2,1,Blue,0055BF,f
3,2,Green,237841,f
4,3,Dark Turquoise,008F9B,f


In [5]:
colors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        190 non-null    int64 
 1   name      190 non-null    object
 2   rgb       190 non-null    object
 3   is_trans  190 non-null    object
dtypes: int64(1), object(3)
memory usage: 6.1+ KB


In [6]:
sets.head()

,set_num,name,year,theme_id,num_parts
0,001-1,Gears,1965,1,43
1,0011-2,Town Mini-Figures,1979,67,12
2,0011-3,Castle 2 for 1 Bonus Offer,1987,199,0
3,0012-1,Space Mini-Figures,1979,143,12
4,0013-1,Space Mini-Figures,1979,143,12


In [7]:
sets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18576 non-null  object
 1   name       18576 non-null  object
 2   year       18576 non-null  int64 
 3   theme_id   18576 non-null  int64 
 4   num_parts  18576 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 725.8+ KB


In [8]:
sets.describe()

,year,theme_id,num_parts
count,18576.000000,18576.000000,18576.000000
mean,2006.200581,405.924365,160.666344
std,13.749798,197.809058,395.885329
min,1949.000000,1.000000,0.000000
25%,2000.000000,246.000000,6.000000
50%,2010.000000,473.000000,37.000000
75%,2017.000000,525.000000,147.000000
max,2022.000000,719.000000,11695.000000


In [9]:
parts.head()

,part_num,name,part_cat_id,part_material
0,003381,Sticker Sheet for Set 663-1,58,Plastic
1,003383,"Sticker Sheet for Sets 618-1, 628-2",58,Plastic
2,003402,"Sticker Sheet for Sets 310-3, 311-1, 312-3",58,Plastic
3,003429,Sticker Sheet for Set 1550-1,58,Plastic
4,003432,"Sticker Sheet for Sets 357-1, 355-1, 940-1",58,Plastic


In [10]:
parts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44781 entries, 0 to 44780
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   part_num       44781 non-null  object
 1   name           44781 non-null  object
 2   part_cat_id    44781 non-null  int64 
 3   part_material  44781 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.4+ MB


In [11]:
themes.head()

,id,name,parent_id
0,1,Technic,NaN
1,3,Competition,1.0
2,4,Expert Builder,1.0
3,16,RoboRiders,1.0
4,17,Speed Slammers,1.0


In [12]:
inventories

,id,version,set_num
0,1,1,7922-1
1,3,1,3931-1
2,4,1,6942-1
3,15,1,5158-1
4,16,1,903-1
...,...,...,...
31582,122328,2,426-1
31583,122329,1,210-2-DUMMY-1
31584,122344,2,7036-1
31585,122450,1,11961-1


In [13]:
inventory_minifigs

,inventory_id,fig_num,quantity
0,3,fig-001549,1
1,4,fig-000764,1
2,19,fig-000555,1
3,25,fig-000574,1
4,26,fig-000842,1
...,...,...,...
18139,122344,fig-009900,1
18140,122344,fig-009908,2
18141,122344,fig-009954,1
18142,122344,fig-009955,1


In [14]:
minifigs

,fig_num,name,num_parts
0,fig-000001,Toy Store Employee,4
1,fig-000002,Customer Kid,4
2,fig-000003,"Assassin Droid, White",8
3,fig-000004,Basic Figure,4
4,fig-000005,Captain America with Short Legs,3
...,...,...,...
11822,fig-012221,Bob-omb,11
11823,fig-012222,Boomer (Banzai) Bill,10
11824,fig-012223,Bullet Bill,10
11825,fig-012224,Buzzy Beetle,7


**Task 1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer). 

In [15]:
def data_count(x):
    return len(x.index)

In [16]:
data_count(sets)

18576

**Task 2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Saved as `average_parts_per_set`.

In [17]:
average_parts_per_set = round(sets['num_parts'].mean())
average_parts_per_set

161

**Task 3:** What is the median number of parts per set (output as an integer)?  Save this as `median_parts_per_set`.

In [18]:
median_parts_per_set = round(sets['num_parts'].median())
median_parts_per_set

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Task 4:** What is the largest number of parts that a set has? Save this as `max_parts_in_set`.

In [19]:
max_parts_in_set = sets['num_parts'].max()
max_parts_in_set

11695

**Task 5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `largest_parts_set_row`.  Do not reset the index.

In [20]:
largest_parts_set_row = sets[sets.num_parts == sets.num_parts.max()]
largest_parts_set_row

,set_num,name,year,theme_id,num_parts
3867,31203-1,World Map,2021,709,11695


**Task 6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [21]:
def select_set_row(x):
    row = sets[sets['set_num']== x]
    return row
select_set_row(3867)

,set_num,name,year,theme_id,num_parts


**Task 7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [22]:
def select_set_numparts(x):
    x = int(x)
    rows = sets[sets['num_parts']== x]
    return rows.sort_values(by = ['year', 'set_num'])

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [23]:
select_set_numparts(0)

,set_num,name,year,theme_id,num_parts
11061,700.1-2,Extra-Large Gift Set (Mursten),1953,366,0
17563,b55dk-01,Lego Mursten - System i Leg Byggebog,1955,497,0
17564,b56de-01,System im Spiel,1956,497,0
11072,700.2-2,Gift Package,1957,366,0
11078,700.3A-2,Gift Package,1957,372,0
...,...,...,...,...,...
15619,854195-1,Ernie Key Chain,2022,503,0
15620,854197-1,Demogorgon Key Chain,2022,503,0
17170,9780744054583-1,DC Comics Super Heroes: Character Encyclopedia...,2022,497,0
17193,9780794447557-1,City: Color The City,2022,497,0


**Task 8:** 
- The oldest set listed was made in what year?  Save the year as `oldest` (as an integer).  
- The newest set listed was made in what year? Save the year as `newest` (as an integer).

In [24]:
oldest = sets['year'].min()
oldest

1949

In [25]:
newest = sets['year'].max()
newest

2022

**Task 9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `oldest` above) and call this `oldest_sets`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `newest` above) and call this `newest_sets`.

In [26]:
def select_set_year(x):
    x = int(x)
    rowy = sets[sets['year']== x]
    return rowy.sort_values(by = ['set_num'])


In [64]:
oldest_sets = select_set_year(1949)
oldest_sets

,set_num,name,year,theme_id,num_parts
11057,700.1-1,Extra-Large Gift Set (ABB),1949,365,142
11070,700.2-1,Large Gift Set (ABB),1949,365,178
11075,700.3-1,Medium Gift Set (ABB),1949,365,142
11080,700.A-1,Small Brick Set (ABB),1949,371,24
11081,700.B-1,Small Doors and Windows Set (ABB),1949,371,12


In [65]:
newest_sets = select_set_year(2022)
newest_sets

,set_num,name,year,theme_id,num_parts
336,10297-1,Boutique Hotel,2022,155,3068
600,10781-1,Spider-Man’s Techno Trike,2022,706,59
601,10782-1,Hulk vs. Rhino Truck Showdown,2022,706,110
602,10783-1,Spider-Man at Doc Ock’s Lab,2022,706,131
603,10784-1,Spider-Man Webquarters Hangout,2022,706,155
...,...,...,...,...,...
17170,9780744054583-1,DC Comics Super Heroes: Character Encyclopedia...,2022,497,0
17193,9780794447557-1,City: Color The City,2022,497,0
17200,9780794448172-1,DC Comics Super Heroes: Ready For Action,2022,497,0
17207,9780794449247-1,DC Comics Super Heroes: Gotham City’s New Defe...,2022,497,7


**Task 10:** What is the median year represented in the `sets` data?  Save this as `median_set_year` (as an integer).

In [29]:
median_set_year = int(sets['year'].median())
median_set_year

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Task 11:** What is the average number of sets listed per year?  Save this as `average_sets_per_year` (as an integer).

In [30]:
average_sets_per_year = sets['year'].value_counts().mean()

average_sets_per_year = int(average_sets_per_year)

average_sets_per_year

258

**Task 12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [31]:


themes['theme_id'] = themes['id']

themes

ty = sets.merge(themes, on= 'theme_id')
ty.rename(columns={'name_y':'name_themes'}, inplace=True)


ty = ty.drop_duplicates(subset=['theme_id', 'year'])


ty

,set_num,name_x,year,theme_id,num_parts,id,name_themes,parent_id
0,001-1,Gears,1965,1,43,1,Technic,NaN
2,1030-1,TECHNIC I: Simple Machines Set,1985,1,211,1,Technic,NaN
4,1039-1,Manual Control Set 1,1986,1,39,1,Technic,NaN
5,1237-1,Honda Promotional Set,2001,1,56,1,Technic,NaN
6,1257-1,Trike Buggy,1999,1,30,1,Technic,NaN
...,...,...,...,...,...,...,...,...
18534,9783-1,Infrared Transmitter with USB Cable,2001,520,1,520,RCX,518.0
18538,9748-1,Droid Developer Kit,1999,261,658,261,Star Wars,258.0
18539,9754-1,Dark Side Developer Kit,2000,261,578,261,Star Wars,258.0
18540,LEGO-Modulex-1,Unused Modulex parts sold by LEGO,1963,716,276,716,Modulex,NaN


In [32]:
def theme_by_year(x):
    x = int(x)
    themes = ty[ty['year']== x]
    themes = themes.sort_values(by= 'id', ).reset_index()
    return themes[['id', 'name_themes']]


**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [33]:
theme_by_year(1960)

,id,name_themes
0,371,Supplemental
1,497,Books
2,513,Classic


**Task 13:** What theme has the most number of sets associated with it?  Save the theme name as `most_common_theme_name` (as a string).

In [34]:
ty2 = sets.merge(themes, on= 'theme_id')
ty2.rename(columns={'name_y':'name_themes', 'name_x':'name_sets'}, inplace=True)


theme_num = ty2['name_themes'].value_counts().index.tolist()
theme_num[0]

most_common_theme_name = theme_num[0]
most_common_theme_name

'Gear'

**Task 14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [35]:
def theme_by_name(name):
    setn = ty2.sort_values(by='set_num')
    setn = setn.loc[setn['name_themes']== name].reset_index()
    return setn[['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']]

**Code Check:** Call the `theme_by_name()` function using the theme name from `most_common_theme_name`.  You should return 1,904 sets that match that theme name.

In [36]:
theme_by_name('Friends')

,set_num,name_sets,year,num_parts,theme_id,name_themes
0,10726-1,Stephanie's Horse Carriage,2016,58,595,Friends
1,10727-1,Emma's Ice Cream Truck,2016,136,595,Friends
2,10728-1,Mia's Vet Clinic,2016,182,595,Friends
3,10746-1,Mia's Farm Suitcase,2017,106,595,Friends
4,10747-1,Andrea and Stephanie's Beach Holiday,2017,144,595,Friends
...,...,...,...,...,...,...
477,FRNDSMAGJNGLE-1,LEGO Friends Magazine Issue 'Jungle adventures...,2014,28,494,Friends
478,ICECREAMTRUCK-1,Ice Cream Truck,2020,60,494,Friends
479,PUPPY-HOUSE-2,Puppy House,2019,72,494,Friends
480,TRUFRIENDS-1,Wishing Well,2016,33,494,Friends


**Task 15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [37]:
def theme_by_setnum(num):
    name_num =ty2.loc[ty2['set_num']== num]['name_themes'].values
    return name_num[0]


**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [38]:
theme_by_setnum('41420-1')

'Friends'

In [39]:
sets.loc[sets['name']== 'Millennium Falcon']

,set_num,name,year,theme_id,num_parts
209,10179-1,Millennium Falcon,2007,171,5197
6570,4488-1,Millennium Falcon,2003,158,87
12450,7190-1,Millennium Falcon,2000,158,671
12719,75030-1,Millennium Falcon,2014,158,94
12790,75105-1,Millennium Falcon,2015,158,1330
12884,75192-1,Millennium Falcon,2017,171,7541
12959,75257-1,Millennium Falcon,2019,158,1351
13846,7965-1,Millennium Falcon,2011,158,1254
16458,911607-1,Millennium Falcon,2016,158,42
16501,911949-1,Millennium Falcon,2019,158,32


**Task 16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You can keep this as an intermediate DataFrame for later steps.

In [40]:

#sets, inventories, inventory_minifigs, and minifigs
si = sets.merge(inventories, on= 'set_num', how='outer')
si.rename(columns= {'id': 'inventory_id'}, inplace = True)
si.shape

sim = si.merge(inventory_minifigs, on= 'inventory_id', how='outer')
sim
simm = sim.merge(minifigs, on='fig_num', how='outer')
simm
simm.rename(columns= {'name_x': 'set_name', 'num_parts_x': 'set_parts', 'name_y': 'fig_name', 'num_parts_y': 'fig_num_parts'}, inplace=True)
simm


,set_num,set_name,year,theme_id,set_parts,inventory_id,version,fig_num,quantity,fig_name,fig_num_parts
0,001-1,Gears,1965.0,1.0,43.0,24696.0,1.0,NaN,NaN,NaN,NaN
1,0011-3,Castle 2 for 1 Bonus Offer,1987.0,199.0,0.0,2216.0,1.0,NaN,NaN,NaN,NaN
2,002-1,4.5V Samsonite Gears Motor Set,1965.0,1.0,3.0,24701.0,1.0,NaN,NaN,NaN,NaN
3,003-1,Master Mechanic Set,1966.0,366.0,403.0,24702.0,1.0,NaN,NaN,NaN,NaN
4,005-1,Basic Building Set in Cardboard,1965.0,366.0,35.0,2028.0,1.0,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
42525,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fig-012221,NaN,Bob-omb,11.0
42526,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fig-012222,NaN,Boomer (Banzai) Bill,10.0
42527,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fig-012223,NaN,Bullet Bill,10.0
42528,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fig-012224,NaN,Buzzy Beetle,7.0


In [41]:
simm.version = simm.version.fillna(0)
simm.version = simm.version.astype(int)
simm


,set_num,set_name,year,theme_id,set_parts,inventory_id,version,fig_num,quantity,fig_name,fig_num_parts
0,001-1,Gears,1965.0,1.0,43.0,24696.0,1,NaN,NaN,NaN,NaN
1,0011-3,Castle 2 for 1 Bonus Offer,1987.0,199.0,0.0,2216.0,1,NaN,NaN,NaN,NaN
2,002-1,4.5V Samsonite Gears Motor Set,1965.0,1.0,3.0,24701.0,1,NaN,NaN,NaN,NaN
3,003-1,Master Mechanic Set,1966.0,366.0,403.0,24702.0,1,NaN,NaN,NaN,NaN
4,005-1,Basic Building Set in Cardboard,1965.0,366.0,35.0,2028.0,1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
42525,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012221,NaN,Bob-omb,11.0
42526,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012222,NaN,Boomer (Banzai) Bill,10.0
42527,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012223,NaN,Bullet Bill,10.0
42528,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012224,NaN,Buzzy Beetle,7.0


In [42]:
simm.fig_num_parts = simm.fig_num_parts.fillna(0)

simm.fig_num_parts = simm.fig_num_parts.astype(int)
simm

,set_num,set_name,year,theme_id,set_parts,inventory_id,version,fig_num,quantity,fig_name,fig_num_parts
0,001-1,Gears,1965.0,1.0,43.0,24696.0,1,NaN,NaN,NaN,0
1,0011-3,Castle 2 for 1 Bonus Offer,1987.0,199.0,0.0,2216.0,1,NaN,NaN,NaN,0
2,002-1,4.5V Samsonite Gears Motor Set,1965.0,1.0,3.0,24701.0,1,NaN,NaN,NaN,0
3,003-1,Master Mechanic Set,1966.0,366.0,403.0,24702.0,1,NaN,NaN,NaN,0
4,005-1,Basic Building Set in Cardboard,1965.0,366.0,35.0,2028.0,1,NaN,NaN,NaN,0
...,...,...,...,...,...,...,...,...,...,...,...
42525,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012221,NaN,Bob-omb,11
42526,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012222,NaN,Boomer (Banzai) Bill,10
42527,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012223,NaN,Bullet Bill,10
42528,NaN,NaN,NaN,NaN,NaN,NaN,0,fig-012224,NaN,Buzzy Beetle,7


In [43]:
simm.dtypes

set_num           object
set_name          object
year             float64
theme_id         float64
set_parts        float64
inventory_id     float64
version            int64
fig_num           object
quantity         float64
fig_name          object
fig_num_parts      int64
dtype: object

In [44]:





#MF1 = simm.loc[simm['set_name']== 'Millennium Falcon']
MF = sets.loc[sets['name']== 'Millennium Falcon']



In [45]:
MF_max = MF.sort_values('num_parts', ascending = False)
MF_max


,set_num,name,year,theme_id,num_parts
12884,75192-1,Millennium Falcon,2017,171,7541
209,10179-1,Millennium Falcon,2007,171,5197
12959,75257-1,Millennium Falcon,2019,158,1351
12790,75105-1,Millennium Falcon,2015,158,1330
13846,7965-1,Millennium Falcon,2011,158,1254
12450,7190-1,Millennium Falcon,2000,158,671
12719,75030-1,Millennium Falcon,2014,158,94
6570,4488-1,Millennium Falcon,2003,158,87
18488,TRUFALCON-1,Millennium Falcon,2016,158,44
16458,911607-1,Millennium Falcon,2016,158,42


**Task 17:** 
- How many minifigs are contained in the set from Task 16?  Save this as `millennium_falcon_minifig_count` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [46]:
inventories.rename(columns= {'id': 'inventory_id'}, inplace = True)

IM = inventories[inventories['set_num']== MF_max.iloc[0]['set_num']]
IM


,inventory_id,version,set_num
11031,19670,1,75192-1


In [47]:
IMF = inventory_minifigs[inventory_minifigs['inventory_id']== IM.iloc[0]['inventory_id']]
IMF
millennium_falcon_minifig_count = IMF['quantity'].sum()
millennium_falcon_minifig_count
IMF

,inventory_id,fig_num,quantity
11492,19670,fig-001714,1
11493,19670,fig-001810,1
11494,19670,fig-002057,1
11495,19670,fig-002514,1
11496,19670,fig-002544,1
11497,19670,fig-002549,1
11498,19670,fig-002550,1
11499,19670,fig-002551,1


In [48]:
inventories

,inventory_id,version,set_num
0,1,1,7922-1
1,3,1,3931-1
2,4,1,6942-1
3,15,1,5158-1
4,16,1,903-1
...,...,...,...
31582,122328,2,426-1
31583,122329,1,210-2-DUMMY-1
31584,122344,2,7036-1
31585,122450,1,11961-1


**Task 18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [49]:
def get_minifigs(x, y):
    y= int(y)
    mf = simm.loc[simm['set_num']== x] 
    mf1 = mf.loc[mf['version']== y]
    mf1 = mf1.sort_values(by = 'fig_num')
    mf1=  mf1[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']].reset_index()
    mf1 = mf1.drop(columns='index') 
    return mf1


get_minifigs('75192-1', 1)

,set_num,fig_num,fig_name,fig_num_parts
0,75192-1,fig-001714,"Chewbacca, Dark Brown",3
1,75192-1,fig-001810,"Finn, Black Undershirt",4
2,75192-1,fig-002057,"Rey, Dark Tan Robe",4
3,75192-1,fig-002514,"C-3PO, Pearl Gold, Colorful Wires, Printed Legs",3
4,75192-1,fig-002544,BB-8,2
5,75192-1,fig-002549,"Princess Leia, Hoth Outfit, White Jacket, Prin...",4
6,75192-1,fig-002550,"Han Solo, Dark Blue Jacket, Dark Brown Legs, P...",4
7,75192-1,fig-002551,"Han Solo, Old, Angry",4


**Task 19:** Using the above function, output information about the minifigs from the set discussed in Task 16.  Save this output as `millennium_falcon_minifigs`.

In [50]:
millennium_falcon_minifigs = get_minifigs('75192-1', 1)
millennium_falcon_minifigs

,set_num,fig_num,fig_name,fig_num_parts
0,75192-1,fig-001714,"Chewbacca, Dark Brown",3
1,75192-1,fig-001810,"Finn, Black Undershirt",4
2,75192-1,fig-002057,"Rey, Dark Tan Robe",4
3,75192-1,fig-002514,"C-3PO, Pearl Gold, Colorful Wires, Printed Legs",3
4,75192-1,fig-002544,BB-8,2
5,75192-1,fig-002549,"Princess Leia, Hoth Outfit, White Jacket, Prin...",4
6,75192-1,fig-002550,"Han Solo, Dark Blue Jacket, Dark Brown Legs, P...",4
7,75192-1,fig-002551,"Han Solo, Old, Angry",4


**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [51]:
get_minifigs('10217-1', 2)

,set_num,fig_num,fig_name,fig_num_parts
0,10217-1,fig-002229,Skeleton (Standard) (3626b Head),6
1,10217-1,fig-006075,"Rubeus Hagrid, Dark Brown Overcoat",3
2,10217-1,fig-006084,"Lucius Malfoy, Black Robes and Cape, Death Eat...",5
3,10217-1,fig-006130,"Harry Potter, Open Dark Blue Jacket over White...",4
4,10217-1,fig-006138,Fenrir Greyback,4
5,10217-1,fig-006162,"Ron Weasley, Red Tartan Sweater Vest, Reddish ...",4
6,10217-1,fig-006163,Fred / George Weasley in Dark Orange Suit,4
7,10217-1,fig-006164,Garrick Ollivander,4
8,10217-1,fig-006165,"Goblin, Black Jacket and Light Bluish Gray Legs",3
9,10217-1,fig-006166,"Goblin, Black Jacket and Reddish Brown Legs",3


**Task 20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [52]:
def minifigs_from_themes(x):
    x = int(x)
    mt = simm.loc[simm['theme_id']== x]
    mt = mt.sort_values(by = 'fig_num')
    mt = mt[['fig_num', 'fig_name']]
    mt = mt.dropna().drop_duplicates()
    mt = mt.reset_index()
    mt = mt.drop(columns='index')
    return mt

minifigs_from_themes(1)

,fig_num,fig_name
0,fig-010488,"Technic Figure Black Legs, Light Gray Top with..."
1,fig-010489,"Technic Figure Black Legs, White Top with Poli..."
2,fig-010490,"Technic Figure Blue Legs, Light Gray Top with ..."
3,fig-010491,"Technic Figure Blue Legs, Light Gray Top with ..."
4,fig-010492,"Technic Figure Red Legs, Red Top with Fire Pri..."
5,fig-010494,"Technic Figure Blue Legs, Red Top with Zip, Bl..."
6,fig-010495,"Technic Figure Yellow Legs, Yellow Top [Power ..."
7,fig-010500,"Technic Figure White Legs, White Top with Red ..."
8,fig-010501,"Technic Figure Blue Legs, White Top with Zip &..."
9,fig-010502,"Technic Figure Blue Legs, White Top with Zip &..."


In [53]:
simm.theme_id = simm.theme_id.fillna(0)
simm.theme_id = simm.theme_id.astype(int)
simm

,set_num,set_name,year,theme_id,set_parts,inventory_id,version,fig_num,quantity,fig_name,fig_num_parts
0,001-1,Gears,1965.0,1,43.0,24696.0,1,NaN,NaN,NaN,0
1,0011-3,Castle 2 for 1 Bonus Offer,1987.0,199,0.0,2216.0,1,NaN,NaN,NaN,0
2,002-1,4.5V Samsonite Gears Motor Set,1965.0,1,3.0,24701.0,1,NaN,NaN,NaN,0
3,003-1,Master Mechanic Set,1966.0,366,403.0,24702.0,1,NaN,NaN,NaN,0
4,005-1,Basic Building Set in Cardboard,1965.0,366,35.0,2028.0,1,NaN,NaN,NaN,0
...,...,...,...,...,...,...,...,...,...,...,...
42525,NaN,NaN,NaN,0,NaN,NaN,0,fig-012221,NaN,Bob-omb,11
42526,NaN,NaN,NaN,0,NaN,NaN,0,fig-012222,NaN,Boomer (Banzai) Bill,10
42527,NaN,NaN,NaN,0,NaN,NaN,0,fig-012223,NaN,Bullet Bill,10
42528,NaN,NaN,NaN,0,NaN,NaN,0,fig-012224,NaN,Buzzy Beetle,7


**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in Task 16.  There should be 149 minifigs output for that respective theme.

In [54]:
minifigs_from_themes(171)

,fig_num,fig_name
0,fig-000003,"Assassin Droid, White"
1,fig-000124,"Chewbacca, Reddish Brown"
2,fig-000144,"Luke Skywalker, White Robe, White Legs, Cheek ..."
3,fig-000303,"Clone Trooper, Phase I Armor, Brown Eyes"
4,fig-000516,"Darth Vader, Light Bluish Gray Skin, Scarred R..."
...,...,...
144,fig-011942,"AT-AT Driver - Light Bluish Grey Helmet, Sand ..."
145,fig-011944,"Snowtrooper, White Hips, Reddish Brown Head, F..."
146,fig-011945,"Snowtrooper, White Hips, Light Flesh Head, Female"
147,fig-011946,"Snowtrooper, White Hips, Light Flesh Head, Off..."


**Task 21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [55]:
def sets_from_minifig(x):
    sfm = simm.loc[simm['fig_num']==x] 
    sfm = sfm.sort_values(by = ['set_num', 'version'])
    sfm['name_sets'] = sfm['set_name']
    sfm = sfm[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']].reset_index()
    sfm = sfm.drop(columns='index')
    return sfm
    


**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [56]:
sets_from_minifig('fig-001039')

,set_num,version,name_sets,fig_num,fig_name
0,30100-1,1,Andrea on the Beach,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."
1,30107-1,1,Birthday Party,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."
2,3061-1,1,City Park Café,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."


**Task 22:** 

- Creates a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [57]:
si
setp = si.merge(inventory_parts, on = 'inventory_id', how = 'outer')
colors['color_id']= colors['id']
setp = setp.merge(colors, on= 'color_id', how = 'outer')
themes['theme_id'] = themes['id']
setp = setp.merge(themes, on='theme_id', how='outer')
setp.rename(columns={'name_x': 'name_sets', 'name_y': 'name_color', 'name': 'name_themes'}, inplace=True)
setp = setp.merge(parts, on = 'part_num', how ='outer')
setp.rename(columns={'name':'name_part'}, inplace=True)
setp
setp = setp.drop(columns=['theme_id', 'num_parts', 'inventory_id', 'color_id', 'id_x', 'rgb', 'is_trans', 'id_y', 'parent_id', 'part_cat_id', 'part_material'])
setp = setp[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color', 'quantity', 'is_spare']]
setp = setp.sort_values(by=['part_num', 'name_color']).reset_index()
setp = setp.drop(columns=['index'])
setp.version = setp.version.fillna(0)
setp.version =setp.version.astype(int)
setp.year = setp.year.fillna(0)
setp.year =setp.year.astype(int)
setp.quantity = setp.quantity.fillna(0)
setp.quantity =setp.quantity.astype(int)
setp

#simm.theme_id = simm.theme_id.fillna(0)
#simm.theme_id = simm.theme_id.astype(int)

,set_num,version,name_sets,year,name_themes,part_num,name_part,name_color,quantity,is_spare
0,663-1,1,Hovercraft,1977,Harbor,003381,Sticker Sheet for Set 663-1,[No Color/Any Color],1,f
1,618-1,1,Police Helicopter,1977,Police,003383,"Sticker Sheet for Sets 618-1, 628-2",[No Color/Any Color],1,f
2,628-2,1,Police Helicopter,1977,Police,003383,"Sticker Sheet for Sets 618-1, 628-2",[No Color/Any Color],1,f
3,310-3,1,Tug,1973,Boat,003402,"Sticker Sheet for Sets 310-3, 311-1, 312-3",[No Color/Any Color],1,f
4,311-1,1,Ferry,1973,Boat,003402,"Sticker Sheet for Sets 310-3, 311-1, 312-3",[No Color/Any Color],1,f
...,...,...,...,...,...,...,...,...,...,...
994370,NaN,0,NaN,0,RC Train,NaN,NaN,NaN,0,NaN
994371,NaN,0,NaN,0,Adventurers,NaN,NaN,NaN,0,NaN
994372,NaN,0,NaN,0,Fabuland,NaN,NaN,NaN,0,NaN
994373,NaN,0,NaN,0,Universal Building Set,NaN,NaN,NaN,0,NaN


In [58]:
def set_parts(x, y):
    y = int(y)
    stp = setp.loc[setp['set_num']==x]
    stp = stp.loc[stp['version']==y]
    stp =stp.reset_index()
    stp =stp.drop(columns=['index'])
    return stp

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in Task 16. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [59]:
set_parts('75192-1', 1)

,set_num,version,name_sets,year,name_themes,part_num,name_part,name_color,quantity,is_spare
0,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,10197,Technic Pin Connector Hub with 2 Perpendicular...,Dark Bluish Gray,3,f
1,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,10201,Bracket 1 x 2 - 1 x 4 [Rounded Corners],Light Bluish Gray,12,f
2,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,10247,Plate Special 2 x 2 with 1 Pin Hole [Complete ...,Dark Bluish Gray,5,f
3,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,10247,Plate Special 2 x 2 with 1 Pin Hole [Complete ...,White,22,f
4,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,10314,"Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...",Tan,3,f
...,...,...,...,...,...,...,...,...,...,...
725,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,99563,Tile Special 1 x 2 with Sloped Walls AKA Money...,Light Bluish Gray,70,f
726,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,99780,Bracket 1 x 2 - 1 x 2 Inverted,Black,6,f
727,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,99780,Bracket 1 x 2 - 1 x 2 Inverted,Dark Bluish Gray,25,f
728,75192-1,1,Millennium Falcon,2017,Ultimate Collector Series,99781,Bracket 1 x 2 - 1 x 2,Light Bluish Gray,2,f


  **Task 23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `unique_color_names`. 

In [60]:
colors1 = colors.copy()
colors1 = colors1.sort_values(by='name')
unique_color_names = colors1.name.unique()
unique_color_names = unique_color_names.tolist()
unique_color_names 


['Aqua',
 'Black',
 'Blue',
 'Blue-Violet',
 'Bright Green',
 'Bright Light Blue',
 'Bright Light Orange',
 'Bright Light Yellow',
 'Bright Pink',
 'Brown',
 'Chrome Antique Brass',
 'Chrome Black',
 'Chrome Blue',
 'Chrome Gold',
 'Chrome Green',
 'Chrome Pink',
 'Chrome Silver',
 'Copper',
 'Coral',
 'Dark Azure',
 'Dark Blue',
 'Dark Blue-Violet',
 'Dark Bluish Gray',
 'Dark Brown',
 'Dark Flesh',
 'Dark Gray',
 'Dark Green',
 'Dark Orange',
 'Dark Pink',
 'Dark Purple',
 'Dark Red',
 'Dark Tan',
 'Dark Turquoise',
 'Earth Orange',
 'Fabuland Brown',
 'Fabuland Orange',
 'Flat Dark Gold',
 'Flat Silver',
 'Flesh',
 'Glitter Trans-Clear',
 'Glitter Trans-Dark Pink',
 'Glitter Trans-Light Blue',
 'Glitter Trans-Neon Green',
 'Glitter Trans-Orange',
 'Glitter Trans-Purple',
 'Glow In Dark Opaque',
 'Glow In Dark Trans',
 'Glow in Dark White',
 'Green',
 'Lavender',
 'Light Aqua',
 'Light Blue',
 'Light Bluish Gray',
 'Light Flesh',
 'Light Gray',
 'Light Green',
 'Light Lime',
 'Light 

**Task 24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `transparent_color_count` (as an integer).

In [61]:
ct = colors[['is_trans']]
ct.replace(['f'], 0, inplace=True)
ct.replace(['t'], 1, inplace=True)
transparent_color_count = ct['is_trans'].sum()
transparent_color_count

/var/folders/hv/6pn1grfs043c6x9_z04cgv5h0000gn/T/ipykernel_83542/4123123314.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ct.replace(['f'], 0, inplace=True)
/var/folders/hv/6pn1grfs043c6x9_z04cgv5h0000gn/T/ipykernel_83542/4123123314.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ct.replace(['t'], 1, inplace=True)


38

**Task 25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [62]:
last_unicorn = inventory_parts.merge(colors, on='color_id', how='outer')
last_unicorn

,inventory_id,part_num,color_id,quantity,is_spare,id,name,rgb,is_trans
0,1.0,48379c01,72,1.0,f,72,Dark Bluish Gray,6C6E68,f
1,22.0,15573,72,1.0,f,72,Dark Bluish Gray,6C6E68,f
2,22.0,2654,72,4.0,f,72,Dark Bluish Gray,6C6E68,f
3,22.0,3070b,72,1.0,t,72,Dark Bluish Gray,6C6E68,f
4,22.0,3070b,72,1.0,f,72,Dark Bluish Gray,6C6E68,f
...,...,...,...,...,...,...,...,...,...
989535,NaN,NaN,1045,NaN,NaN,1045,Modulex Foil Light Blue,68AECE,f
989536,NaN,NaN,1046,NaN,NaN,1046,Modulex Foil Violet,4B0082,f
989537,NaN,NaN,1047,NaN,NaN,1047,Modulex Foil Red,8B0000,f
989538,NaN,NaN,1048,NaN,NaN,1048,Modulex Foil Yellow,FED557,f


In [63]:
def get_part_colors(x):
    gpc = last_unicorn.loc[last_unicorn['part_num']==x]
    gpc1 = gpc.drop(columns=['inventory_id', 'color_id', 'quantity', 'is_spare'])
    gpc1 = gpc1.sort_values(by = 'id')
    gpc1 = gpc1.drop_duplicates(subset=['name'])
    gpc1 = gpc1.reset_index()
    gpc1 = gpc1.drop(columns= ['index' , 'part_num'])
    return gpc1
get_part_colors('10247')

,id,name,rgb,is_trans
0,0,Black,05131D,f
1,4,Red,C91A09,f
2,14,Yellow,F2CD37,f
3,15,White,FFFFFF,f
4,19,Tan,E4CD9E,f
5,71,Light Bluish Gray,A0A5A9,f
6,72,Dark Bluish Gray,6C6E68,f
